diff --git a/lib/stream.js b/lib/stream.js
index 1f27f447995ffbc37739c8750e83dd627881b52f..19d946b791fcdd400115935a999e406a2c96cdb8 100644
--- a/lib/stream.js
+++ b/lib/stream.js
@@ -457,20 +457,19 @@ export class HttpWTStream {
           const res = this.abortres
           this.abortres = null
           res()
-          if (this.readable)
-            this.parentobj.removeReceiveStream(
-              this.readable,
-              this.readableController
-            )
-          if (this.writable)
-            this.parentobj.removeSendStream(
-              this.writable,
-              this.writableController
-            )
-          this.readableclosed = true
-          this.parentobj.removeStreamObj(this)
         }
-
+        if (this.readable)
+          this.parentobj.removeReceiveStream(
+            this.readable,
+            this.readableController
+          )
+        if (this.writable)
+          this.parentobj.removeSendStream(
+            this.writable,
+            this.writableController
+          )
+        this.readableclosed = true
+        this.parentobj.removeStreamObj(this)
         break
 
       case 'streamFinal':
@@ -482,6 +481,7 @@ export class HttpWTStream {
             res()
           }
         }
+        this.parentobj.removeStreamObj(this)
         break
       default:
         log.error('onStreamNetworkFinish unknown task', args.nettask)
